# I. Trắc nghiệm (5 câu, mỗi câu 1 điểm)

Câu trả lời được lựa chọn sẽ được tô đậm

1) Hiện tượng Overfitting là gì ?

**A. Mô hình dự báo đúng trên tập train nhưng không dự báo đúng trên tập test.**

B. Mô hình dự báo không đúng trên cả tập train và tập test.

C. Mô hình dự báo đúng trên tập test nhưng không đúng trên tập train.

D. Mô hình dự báo đúng cả trên tập train và tập test.

2) Để kiểm soát hiện tượng overfitting thì mô hình hồi qui tuyến tính thực hiện điều gì trong lớp mô hình Ridge Regression, Lasso Regression và ElasticNet Regression?

A. Loại bỏ bớt biến đầu vào.

**B. Cộng thêm _thành phần điều chuẩn_ (Regularized Term) vào hàm mất mát để kiểm soát độ lớn của hệ số ước lượng.**

C. Thực hiện Polynormial Feature để tạo ra các biến đầu vào mới là biến đa thức.

D. Thay đổi hàm mục tiêu.

3) Sự khác biệt chính về nghiệm giữa Ridge Regression và Lasso Regression là gì?

A. Ridge Regression thường trả về các trọng số ước lượng có giá trị bằng 0 trong khi các hệ số được ước lượng từ Lasso Regression thường có giá trị nhỏ gần bằng 0.

**B. Lasso Regression thường trả về các trọng số ước lượng có giá trị bằng 0 trong khi các hệ số được ước lượng từ Ridge Regression thường có giá trị nhỏ gần bằng 0.**

C. Các hệ số của Ridge Regression có giá trị lớn trong khi Lasso Regression có giá trị nhỏ.

D. Các hệ số của Lasso Regression có giá trị lớn trong khi Ridge Regression có giá trị nhỏ.

4) Sự khác biệt chính giữa Ridge Regression và Lasso Regression là gì?

**A. Ridge Regression có _thành phần điều chuẩn_ là norm chuẩn bậc 2 trong khi Lasso Regression có là norm chuẩn bậc 1.**

B. Lasso Regression có _thành phần điều chuẩn_ là norm chuẩn bậc 2 trong khi Ridge Regression có là norm chuẩn bậc 1.

C. Ridge Regression luôn đảm bảo có nghiệm trong khi Lasso Regression không thể tìm được công thức gnhiệm trực tiếp từ đại số tuyến tính.

D. Đáp án A và C.

5) Hàm mục tiêu của ElasticNet Regression là gì? Biết rằng trong công thức bên dưới thì $L_1$ và $L_2$ lần lượt là norm chuẩn bậc 1 và bậc 2: 

A. $\text{Regularized Loss} = \text{MSE Loss} + \alpha L_1(\mathbf{w})$

B. $\text{Regularized Loss} = \text{MSE Loss} + \alpha L_2(\mathbf{w})$

C. $\text{Regularized Loss} = \text{MSE Loss} + \alpha L_1(\mathbf{w}) L_2(\mathbf{w})$

**D. $\text{Regularized Loss} = \text{MSE Loss} + \alpha [\lambda L_1(\mathbf{w}) + \frac{(1-\lambda)}{2} L_2(\mathbf{w})]$**

# II. Thực hành (1 câu, 5 điểm)

Xây dựng mô hình hồi qui tuyến tính dựa vào dữ liệu từ một trong các cuộc thi sau đây:

**Stock price forecast**
* https://www.kaggle.com/c/the-winton-stock-market-challenge
* https://www.kaggle.com/c/two-sigma-financial-news
* https://www.kaggle.com/c/jane-street-market-prediction
* https://www.kaggle.com/faressayah/stock-market-analysis-prediction-using-lstm

**Weather forecast (humidity, temperature, wind speed)**
* https://www.kaggle.com/c/rainfall-prediction-challenge
* https://www.kaggle.com/jsphyg/weather-dataset-rattle-package
* https://www.kaggle.com/c/dmbi-18-metro
* https://www.kaggle.com/c/1056lab-temperature-forecasting/
* https://www.kaggle.com/c/dmbi-18-metro/data

**Total consumption of customer**
* https://www.kaggle.com/c/ga-customer-revenue-prediction
* https://www.kaggle.com/c/santander-customer-transaction-prediction

**Churn customer**
* https://www.kaggle.com/c/customer-churn-prediction-2020
* https://www.kaggle.com/c/predict-the-churn-for-customer-dataset
* https://www.kaggle.com/c/1056lab-credit-card-customer-churn-prediction

**Income forecast**
* https://www.kaggle.com/c/ga-customer-revenue-prediction
* https://www.kaggle.com/c/income-level-prediction
* https://www.kaggle.com/c/ml-l1-finalcs
* https://www.kaggle.com/c/tcdml1920-income-ind

**Inventory** 
* https://www.kaggle.com/c/dsedelhi
* https://www.kaggle.com/c/inventory-optimization
* https://www.kaggle.com/c/demand-forecasting-kernels-only
* https://www.kaggle.com/c/grupo-bimbo-inventory-demand

**Taxi demand prediction**
* https://www.kaggle.com/sohaibanwaar1203/taxi-demand-prediction
* https://www.kaggle.com/c/new-york-city-taxi-fare-prediction


Bộ dữ liệu lựa chọn: https://www.kaggle.com/competitions/the-winton-stock-market-challenge/data

In [2]:
import numpy as np 
import pandas as pd 
import os
from google.colab import drive


In [3]:
drive.mount('/content/drive')
os.chdir('drive/MyDrive/Colab Notebooks/Buổi 5')

Mounted at /content/drive


In [43]:
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test_2.csv')
print(df_train.shape)
print(df_test.shape)

(40000, 211)
(120000, 147)


In [5]:
df_train.head()

Id  Feature_1  Feature_2  Feature_3  Feature_4  Feature_5  Feature_6  \
0   1        NaN        NaN        NaN        NaN        8.0        NaN   
1   2        NaN        NaN        NaN        NaN        3.0   0.388896   
2   3        NaN  -0.696727   0.739591  -0.167928        9.0   0.471947   
3   4        NaN  -0.694350   1.568248   0.479073        5.0   0.120653   
4   5        6.0  -1.736489   2.765531   1.245280        7.0   4.866985   

   Feature_7  Feature_8  Feature_9  ...   Ret_175   Ret_176   Ret_177  \
0      75751     0.2254       11.0  ... -0.002688  0.002246 -0.000838   
1      17369     0.0166       13.0  ... -0.000129  0.000123  0.000248   
2       8277     0.3650        9.0  ... -0.000524 -0.000394  0.000116   
3      22508     0.2654       13.0  ...  0.000346 -0.000090  0.000288   
4      22423     0.2138       13.0  ... -0.001235  0.000027  0.002449   

        Ret_178   Ret_179   Ret_180  Ret_PlusOne  Ret_PlusTwo  \
0 -6.953224e-04  0.000003 -0.001974    -0.019512     0.028846   
1  3.315418e-07  0.000003  0.000027    -0.002939    -0.010253   
2  5.322557e-04  0.000274  0.000784    -0.024791     0.015711   
3 -1.281102e-04  0.000074  0.000341    -0.005680    -0.002190   
4  8.619882e-06  0.001209 -0.000004     0.036104    -0.026552   

   Weight_Intraday  Weight_Daily  
0     1.251508e+06  1.564385e+06  
1     1.733950e+06  2.167438e+06  
2     1.529197e+06  1.911497e+06  
3     1.711569e+06  2.139462e+06  
4     1.267270e+06  1.584088e+06  

[5 rows x 211 columns]

Đối với file train, có 211 cột, file test có 147 cột: 
- 1 cột Id
- 25 cột Feature_1 tới Feature_25: các features khác có liên quan tới việc dự báo
- 1 cột Ret_MinusTwo: chênh lệch về giá đóng phiên của ngày D-2 so với ngày D-1 (1 ngày)
- 1 cột Ret_MinusOne: chênh lệch về giá đóng phiên của ngày D-1 so với thời điểm giao dịch trong ngày bắt đầu (1/2 ngày)
- 179 cột Ret_2 tới Ret_180: chênh lệch về giá giữa một phút và phút trước đó trong ngày giao dịch. (đối với file test thì bỏ 60 cột từ Ret 121 tới 180).
- 1 cột Ret_PlusOne: chênh lệch về giá giữa thời điểm Ret_180 của ngày D với thời điểm đóng phiên ngày D+1 (1 ngày, đối với file test bỏ cột này)
- 1 cột Ret_PlusTwo: chênh lệch về giá giữa thời điểm đóng phiên ngày D+1 với thời điểm đóng phiên ngày D+2 (1 ngày, đối với file test bỏ cột này)
- 1 cột Weight_Intraday: trọng số dùng để đánh giá dự đoán trong ngày, các dự đoán từ phút 121 tới 180 (file test bỏ cột này)
- 1 cột Weight_Daily: trọng số dùng để đánh giá dự đoán ngày tiếp theo, các dự đoán ngày D+1 và D+2 (file test bỏ cột này)

In [17]:
features_variables = ['Feature_{}'.format(i) for i in range(1,26)] #các biến có thể dùng hoặc không dùng
return_variables_X = ['Ret_MinusTwo','Ret_MinusOne']+['Ret_{}'.format(i) for i in range(2,121)] #các biến features
return_variables_Y = ['Ret_{}'.format(i) for i in range(121,181)] + ['Ret_PlusOne','Ret_PlusTwo'] #các biến target

In [47]:
#thay các giá trị missing trong các DÒNG bằng trung bình của DÒNG đó, vì mỗi dòng là một cổ phiếu khác nhau

TRAINING_DATA = df_train[return_variables_X+return_variables_Y]
TRAINING_DATA = TRAINING_DATA.T.fillna(TRAINING_DATA.mean(axis=1)).T

TESTING_DATA = df_test[return_variables_X]
TESTING_DATA = TESTING_DATA.T.fillna(TESTING_DATA.mean(axis=1)).T


In [41]:
from sklearn.linear_model import LinearRegression 
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from tqdm.notebook import tqdm

## Dự đoán không sử dụng 25 cột feature

Do các giá trị dự đoán là rất nhỏ (< 1) nên sử dụng đánh giá là MAPE

In [9]:
lr = []
for i in range(62):
  lr = lr + [LinearRegression()]

### Dự đoán phút thứ 121

lr[0] sẽ tương ứng với Ret_121 <br>
lr[1] tương ứng Ret_122 <br>
... <br>
lr[59] tương ứng Ret_180 <br>
lr[60] tương ứng Ret_PlusOne <br>
lr[61] tương ứng Ret_PlusTwo <br>

In [48]:
X_train = TRAINING_DATA[return_variables_X]
Y_train = TRAINING_DATA['Ret_121']

x_train, x_val, y_train, y_val = train_test_split(X_train, Y_train, random_state = 42, test_size = 0.2)

lr[0].fit(x_train, y_train)
print('MAPE of train set:',mean_absolute_percentage_error(lr[0].predict(x_train), y_train))
print('MAPE of validation set:',mean_absolute_percentage_error(lr[0].predict(x_val), y_val))

MAPE of train set: 18.195097889047865
MAPE of validation set: 15.536756737000768


### Dự đoán tất cả các kết quả nhưng chỉ dựa vào dữ liệu có sẵn

Xây dựng 61 model dựa đoán các giá trị từ ret_121 đến ret_180 và ret_plusone, ret_plustwo nhưng chỉ dựa vào 121 biến ban đầu

In [49]:
X_train = TRAINING_DATA[return_variables_X] #X_train không thay đổi 
mape_train_1 = []
mape_val_1 = []

for i in range(len(return_variables_Y)):
  Y_train = TRAINING_DATA[return_variables_Y[i]]
  x_train, x_val, y_train, y_val = train_test_split(X_train, Y_train, random_state = 42, test_size = 0.2)
  lr[i].fit(x_train, y_train)
  mape_train_1.append(mean_absolute_percentage_error(lr[i].predict(x_train), y_train))
  mape_val_1.append(mean_absolute_percentage_error(lr[i].predict(x_val), y_val))

print('Average MAPE for training set:',np.mean(mape_train_1))
print('Average MAPE for validation set:',np.mean(mape_val_1))

Average MAPE for training set: 38.22659922922804
Average MAPE for validation set: 28.854154281932818


### Dự đoán các thời điểm sau này với dữ liệu từ dữ đoán của thời điểm trước

Xây dựng 61 model dự đoán các giá trị, với biến đầu vào của mô hình sau sẽ gồm dữ liệu đầu vào đã cho và kết quả dự đoán của các mô hình đi trước

In [50]:
X_train = TRAINING_DATA[return_variables_X] #X_train ban đầu 
mape_train_2 = []
mape_val_2 = []

for i in range(len(return_variables_Y)):
  Y_train = TRAINING_DATA[return_variables_Y[i]]
  x_train, x_val, y_train, y_val = train_test_split(X_train, Y_train, random_state = 42, test_size = 0.2)
  lr[i].fit(x_train, y_train)
  mape_train_2.append(mean_absolute_percentage_error(lr[i].predict(x_train), y_train))
  mape_val_2.append(mean_absolute_percentage_error(lr[i].predict(x_val), y_val))
  X_train.insert(121+i, return_variables_Y[i], lr[i].predict(X_train))  #thêm kết quả dự đoán vào X_train, chuẩn bị cho mô hình tiếp theo

print('Average MAPE for training set:',np.mean(mape_train_2))
print('Average MAPE for validation set:',np.mean(mape_val_2))

Average MAPE for training set: 24.329132654988346
Average MAPE for validation set: 155723.4881580556


Có thể thấy cách thức dự đáon với dữ liệu từ thời điểm trước cho ra kết quả tốt hơn (MAPE nhỏ hơn). Với mô hình xây dựng theo cách 2, kết quả bị overfit rất nặng. 

Khi thử trên dữ liệu đầy đủ (tất cả các cột đều không có missing value), kết quả dự đoán tốt hơn, chỉ dừng ở mức 10%

### Dự đoán test set với mô hình xây dựng theo cách đầu tiên

In [52]:
#Xây dựng mô hình
X_train = TRAINING_DATA[return_variables_X] 
X_test = TESTING_DATA[return_variables_X]
Y_hat = {}
#Y_hat là ma trận kết quả
for i in range(len(return_variables_Y)):
  Y_train = TRAINING_DATA[return_variables_Y[i]]
  lr[i].fit(X_train, Y_train)
  Y_hat[return_variables_Y[i]] = lr[i].predict(X_test)

Y_hat = pd.DataFrame(Y_hat)
print('5 dòng đầu của ma trận kết quả dự đoán')
Y_hat.head()

5 dòng đầu của ma trận kết quả dự đoán


Ret_121   Ret_122   Ret_123   Ret_124   Ret_125   Ret_126   Ret_127  \
0  0.000152  0.000272  0.000449 -0.000098  0.000005 -0.000178  0.000044   
1 -0.000291  0.000339 -0.000153 -0.000172 -0.000026  0.000048  0.000101   
2 -0.000226  0.000076 -0.000100  0.000020  0.000083 -0.000208  0.000292   
3  0.000201  0.000201  0.000129 -0.000025  0.000018  0.000006  0.000142   
4 -0.000047  0.000107  0.000044  0.000034 -0.000080 -0.000093  0.000077   

    Ret_128   Ret_129   Ret_130  ...   Ret_173   Ret_174   Ret_175   Ret_176  \
0 -0.000466  0.000085 -0.000137  ... -0.000121 -0.000039 -0.000007  0.000071   
1 -0.000124 -0.000132  0.000027  ...  0.000018  0.000167 -0.000033 -0.000110   
2  0.000240  0.000067  0.000053  ... -0.000218 -0.000006  0.000029 -0.000038   
3  0.000381 -0.000204  0.000377  ... -0.000246  0.000273  0.000106 -0.000086   
4 -0.000135  0.000003  0.000031  ...  0.000037 -0.000029  0.000011 -0.000082   

    Ret_177   Ret_178   Ret_179   Ret_180  Ret_PlusOne  Ret_PlusTwo  
0  0.000045  0.000072  0.000125  0.000203    -0.006665    -0.002842  
1 -0.000061 -0.000082  0.000024 -0.000030     0.000786    -0.006752  
2 -0.000017  0.000049  0.000020  0.000343     0.004198     0.001780  
3  0.000066 -0.000060 -0.000034  0.000307    -0.002594    -0.000451  
4 -0.000088 -0.000060 -0.000133 -0.000047    -0.002636     0.000178  

[5 rows x 62 columns]